onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlSTFT.h

```diff
@@ -0,0 +1,610 @@
+        ComPtr<ID3D12Resource> persistentResource;
+        ComPtr<IUnknown> persistentResourcePoolingUnk;
+        std::optional<DML_BUFFER_BINDING> persistentResourceBinding;
+        bool hasWindowTensor = false;
+        uint64_t signalBufferSizeInBytes = 0;
+        uint64_t windowBufferSizeInBytes = 0;
+        uint64_t outputBufferSizeInBytes = 0;
+    } m_framingOperator;
+
+    struct
+    {
+        ComPtr<GpuDFTOperator> op;
+        std::array<uint32_t, 3> inputDims;
+        std::array<uint32_t, 3> outputDims;
+        uint32_t dftLength = 0;
+    } m_dftOperator;
+
+public:
+    DmlSTFTOperator(IMLOperatorKernelCreationContext* context)
+    {
+        ComPtr<IMLOperatorKernelCreationContextNodeWrapperPrivate> contextPrivate;
+        ORT_THROW_IF_FAILED(context->QueryInterface(IID_PPV_ARGS(&contextPrivate)));
+
+        ComPtr<IUnknown> provider;
+        ORT_THROW_IF_FAILED(contextPrivate->GetExecutionProvider(&provider));
+        ORT_THROW_IF_FAILED(provider.As(&m_dmlProvider));
+        ORT_THROW_IF_FAILED(m_dmlProvider->GetDmlDevice(&m_dmlDevice));
+        ORT_THROW_IF_FAILED(m_dmlProvider->GetD3DDevice(&m_d3dDevice));
+
+        ComPtr<IMLOperatorTensorShapeDescription> shapeDescInfo;
+        ORT_THROW_IF_FAILED(context->GetTensorShapeDescription(&shapeDescInfo));
+
+        DmlSTFTConstructionParams params(contextPrivate.Get(), shapeDescInfo.Get());
+
+        CompileAndInitFramingOperator(params);
+
+        constexpr uint32_t dftAxis = 1;
+        constexpr bool dftIsInverse = false;
+        m_dftOperator.op = wil::MakeOrThrow<GpuDFTOperator>(
+            m_d3dDevice.Get(),
+            dftAxis,
+            params.isOnesided,
+            dftIsInverse,
+            Dml::GetMlDataTypeFromDmlDataType(params.dataType)
+        );
+
+        m_dftOperator.inputDims = { params.batchSize * params.frameCount, params.frameSize, 1 };
+        m_dftOperator.outputDims = { params.batchSize * params.frameCount, params.frameDftElementCount, 2 };
+        m_dftOperator.dftLength = params.frameSize;
+    }
+
+    void CompileAndInitFramingOperator(const DmlSTFTConstructionParams& params)
+    {
+        StackAllocator<1024> stackAllocator;
+
+        auto FillTensorDesc = [&params, &stackAllocator](
+            DML_TENSOR_DESC* tensorDesc,
+            uint64_t* bufferSizeInBytes,
+            std::initializer_list<uint32_t> strides = {}
+            )
+        {
+            DmlBufferTensorDesc bufferDesc;
+            bufferDesc.dataType = params.dataType;
+            bufferDesc.sizes = { params.batchSize, params.frameCount, params.frameSize };
+            if (strides.size() > 0) { bufferDesc.strides = strides; }
+            bufferDesc.totalTensorSizeInBytes = DMLCalcBufferTensorSize(
+                bufferDesc.dataType,
+                onnxruntime::narrow<uint32_t>(bufferDesc.sizes.size()),
+                bufferDesc.sizes.data(),
+                strides.size() > 0 ? bufferDesc.strides->data() : nullptr
```

Comment:
>strides.size() > 0 ? bufferDesc.strides->data() : nullptr [](http://example.com/codeflow?start=16&length=57)

[minor] Wouldn't just `strides.data()` work here? If strides is empty, then it will be nullptr anyway (a default constructed initializer list is null with 0 count).

Reply:
`std::initializer_list` only has a `std::data` overload in C++17 (no member function `.data()`). Even so, `bufferDesc.strides` is not a raw pointer but a `std::optional<std::vector<uint32_t>>`.

Reply:
Yeah, `bufferDesc.strides` is an optional, is  which why I suggested using `strides.data()`, not `bufferDesc.strides.data()`. Yeah, you're right about `.data()` - what a silly gap in the initializer_list API, that it has begin/end/size and is missing data. So `std::data(strides)` is what I should have said. :b